src/credentials/GroupDataProvider.h

```diff
@@ -0,0 +1,154 @@
+        {
+            if (this == &other)
+                return *this; // Guard self assignment
+            this->endpoint = other.endpoint;
+            this->group    = other.group;
+            return *this;
+        }
+    };
+
+    struct GroupState
+    {
+        uint16_t state_index;
+        chip::GroupId group;
+        uint16_t key_set_index;
+        bool operator==(const GroupState & other)
+        {
+            return this->state_index == other.state_index && this->group == other.group &&
+                this->key_set_index == other.key_set_index;
+        }
+    };
+
+    struct KeySet
+    {
+        enum class SecurityPolicy : uint8_t
+        {
+            Standard   = 0,
+            LowLatency = 1
+        };
+
+        uint16_t key_set_index;
+        SecurityPolicy policy;
+        EpochKey epoch_keys[3];
+    };
+
+    template <typename T>
+    class Iterator
+    {
+    public:
+        virtual ~Iterator()      = default;
+        virtual uint16_t Count() = 0;
+        virtual bool HasNext()   = 0;
+        virtual T Next()         = 0;
+
+    protected:
+        Iterator() = default;
+    };
+
+    struct GroupListener
+    {
+        void OnStateChanged(GroupState & old_state, GroupState & new_state);
+    };
+
+    GroupDataProvider()          = default;
+    virtual ~GroupDataProvider() = default;
+
+    // Not copyable
+    GroupDataProvider(const GroupDataProvider &) = delete;
+    GroupDataProvider & operator=(const GroupDataProvider &) = delete;
+
+    virtual CHIP_ERROR Init() = 0;
+    virtual void Finish()     = 0;
+
+    // Endpoints
+    virtual bool Exists(chip::FabricIndex fabric, GroupMapping & mapping)                       = 0;
+    virtual CHIP_ERROR Add(chip::FabricIndex fabric, GroupMapping & mapping, const char * name) = 0;
+    virtual CHIP_ERROR Remove(chip::FabricIndex fabric, GroupMapping & mapping)                 = 0;
+    virtual CHIP_ERROR RemoveEndpoint(chip::FabricIndex fabric, EndpointId endpoint)            = 0;
+    virtual Iterator<GroupId> * IterateEndpoint(chip::FabricIndex fabric, EndpointId endpoint)  = 0;
+
+    // States
+    virtual CHIP_ERROR Set(chip::FabricIndex fabric, GroupState & state)           = 0;
+    virtual CHIP_ERROR Get(chip::FabricIndex fabric, GroupState & state)           = 0;
+    virtual CHIP_ERROR RemoveState(chip::FabricIndex fabric, uint16_t state_index) = 0;
+    virtual Iterator<GroupState *> * IterateStates(chip::FabricIndex fabric)       = 0;
+
+    // Keys
+    virtual CHIP_ERROR Set(chip::FabricIndex fabric, KeySet & keys)                 = 0;
+    virtual CHIP_ERROR Get(chip::FabricIndex fabric, KeySet & keys)                 = 0;
+    virtual CHIP_ERROR RemoveKeys(chip::FabricIndex fabric, uint16_t key_set_index) = 0;
+    virtual Iterator<KeySet *> * IterateKeys(chip::FabricIndex fabric)              = 0;
+
+    void SetListener(GroupListener * listener) { mListener = listener; };
+    void RemoveListener() { mListener = nullptr; };
+
```

Comment:
During discussion about this API, we had agreed to:

```
    // From an incoming group message, iterate through possible keys, filtered by
    // group_id (if known) and group_session_id (key hash filter).
    // TODO: Conceptually easy, but needs to be a “DECRYPT iterator” so that the key is not
    // extricated from SE if using SE/HSM/Trustzone secure key storage
    virtual GroupKeyCandidateIterator * IterateGroupKeyCandidates(bool group_id_present, uint16_t group_id, uint16_t group_session_id) = 0;
```

and

```
  // Encrypt for group transmit
    // TODO: Refactor possibly as EncryptForGroup(chip::FabricIndex fabric_index, uint16_t group_id, …)
    virtual void GetEpochKeyForGroupTransmit(chip::FabricIndex fabric_index, uint16_t group_id, EpochKey & out_epoch_key) = 0;
```

In other words, ways to use the group keys so that they could be stored in secure element/HSM rather than requiring direct key access.

Reply:
Let's punt on this API as a refactor. @jepenven-silabs please open an issue to avoid iterating over the raw keys.